File: ios/Classes/XfyunvoicePlugin.m
#import "XfyunvoicePlugin.h"

@interface XfyunvoicePlugin ()<IFlySpeechRecognizerDelegate,IFlySpeechSynthesizerDelegate>{
    NSString *_result; // 记录语音识别返回的结果
}


@end

@implementation XfyunvoicePlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel *channel = [FlutterMethodChannel methodChannelWithName:@"xfyunvoice" binaryMessenger:[registrar messenger]];
    XfyunvoicePlugin* instance = [[XfyunvoicePlugin alloc] init];
    [instance setChannelMethod:channel];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"setAppId" isEqualToString:call.method]) {
      NSDictionary *data = call.arguments;
      if([data objectForKey:@"args"]){
          NSString *appId = [NSString stringWithFormat: @"%@",[data objectForKey:@"args"]];
          NSLog(@"=======>appId:%@",appId);
          [IFlySpeechUtility createUtility:appId];
          if (self.speechRecognizer == nil) {
              [self initRecognizer];
          }
          result(@YES);
      }else{
          result(@NO);
      }
  }else if([@"requestPermission" isEqualToString: call.method]){
      NSURL *url = [NSURL URLWithString:UIApplicationOpenSettingsURLString];
      if( [[UIApplication sharedApplication]canOpenURL:url] ) {
          [[UIApplication sharedApplication]openURL:url];
      }
  }else if([@"startListening" isEqualToString:call.method]){
      AVAuthorizationStatus authStatus = [AVCaptureDevice authorizationStatusForMediaType:AVMediaTypeAudio];
      if(authStatus == AVAuthorizationStatusRestricted || authStatus == AVAuthorizationStatusDenied) {
          result(@{@"ret":@NO,@"msg":@"App未获得录音权限",@"code":@(-1)});
      }else{
          if (self.speechRecognizer == nil) {
              [self initRecognizer];
          }
          [_speechRecognizer cancel];
          if(_speechSynthesizer != nil){
              [_speechSynthesizer stopSpeaking];
          }
          [_speechRecognizer setParameter:IFLY_AUDIO_SOURCE_MIC forKey:@"audio_source"];
          [_speechRecognizer setParameter:@"json" forKey:[IFlySpeechConstant RESULT_TYPE]];
          [_speechRecognizer setParameter:@"asr.pcm" forKey:[IFlySpeechConstant ASR_AUDIO_PATH]];
          [_speechRecognizer setDelegate:self];
          BOOL ret = [_speechRecognizer startListening];
          
          result(@{@"ret":@(ret),@"msg":@"正在语音识别中",@"code":@0});
      }
  }else if([@"stopListening" isEqualToString:call.method]){
      if(_speechRecognizer != nil){
          [self.speechRecognizer stopListening];
      }
  }else if([@"cancel" isEqualToString:call.method]){
      if(_speechRecognizer != nil){
          [self.speechRecognizer cancel];
      }
  }else if([@"startSpeaking" isEqualToString:call.method]){
      NSString *msg = [NSString stringWithFormat: @"%@",call.arguments];
      if(_speechSynthesizer == nil){
          [self initSynthesizer];
      }
      [_speechSynthesizer stopSpeaking];
      [_speechSynthesizer setDelegate:self];
      [_speechSynthesizer startSpeaking: msg];
      result(@(YES));
  }else if([@"isSpeaking" isEqualToString:call.method]){
      if(_speechSynthesizer == nil){
          result(@(NO));
      }else{
          result(@([_speechSynthesizer isSpeaking]));
      }
  }else if([@"pauseSpeaking" isEqualToString:call.method]){
      if(_speechSynthesizer != nil){
          [_speechSynthesizer pauseSpeaking];
      }
  }else if([@"resumeSpeaking" isEqualToString:call.method]){
      if(_speechSynthesizer != nil){
          [_speechSynthesizer resumeSpeaking];
      }
  }else if([@"destroy" isEqualToString:call.method]){
      if(_speechSynthesizer == nil){
          result(@(NO));
      }else{
          result(@([IFlySpeechSynthesizer destroy]));
      }
  }else {
    result(FlutterMethodNotImplemented);
  }
}

-(void)initSynthesizer{
    if(_speechSynthesizer == nil){
        _speechSynthesizer = [IFlySpeechSynthesizer sharedInstance];
    }
    [[IFlySpeechUtility getUtility] setParameter:@"tts" forKey:[IFlyResourceUtil ENGINE_START]];
    
    [_speechSynthesizer setParameter:@"50" forKey:[IFlySpeechConstant SPEED]];
    [_speechSynthesizer setParameter:@"50" forKey:[IFlySpeechConstant VOLUME]];
    [_speechSynthesizer setParameter:@"50" forKey:[IFlySpeechConstant PITCH]];
    [_speechSynthesizer setParameter:@"16000" forKey:[IFlySpeechConstant SAMPLE_RATE]];
    [_speechSynthesizer setParameter:@"xiaoyan" forKey:[IFlySpeechConstant VOICE_NAME]];
    [_speechSynthesizer setParameter:@"unicode" forKey:[IFlySpeechConstant TEXT_ENCODING]];
    [_speechSynthesizer setParameter:@"cloud" forKey:[IFlySpeechConstant ENGINE_TYPE]];
}

-(void)initRecognizer{
    if (self.speechRecognizer == nil) {
        [self setSpeechRecognizer:[IFlySpeechRecognizer sharedInstance]];
    }
    [self.speechRecognizer setParameter:@"" forKey:[IFlySpeechConstant PARAMS]];
    [self.speechRecognizer setParameter:@"iat" forKey:[IFlySpeechConstant IFLY_DOMAIN]];
    if (_speechRecognizer != nil) {
        [_speechRecognizer setParameter:@"30000" forKey:[IFlySpeechConstant SPEECH_TIMEOUT]];
        [_speechRecognizer setParameter:@"2000" forKey:[IFlySpeechConstant VAD_EOS]];
        [_speechRecognizer setParameter:@"3000" forKey:[IFlySpeechConstant VAD_BOS]];
        [_speechRecognizer setParameter:@"10000" forKey:[IFlySpeechConstant NET_TIMEOUT]];
        [_speechRecognizer setParameter:@"16000" forKey:[IFlySpeechConstant SAMPLE_RATE]];
        [_speechRecognizer setParameter:@"zh_cn" forKey:[IFlySpeechConstant LANGUAGE]];
        [_speechRecognizer setParameter:@"mandarin" forKey:[IFlySpeechConstant ACCENT]];
        [_speechRecognizer setParameter:@"0" forKey:[IFlySpeechConstant ASR_PTT]];
        
        [_speechRecognizer setParameter:@"json" forKey:[IFlySpeechConstant RESULT_TYPE]];
        [_speechRecognizer setParameter:IFLY_AUDIO_SOURCE_STREAM forKey:@"audio_source"];
    }
}



- (void)onCompleted:(IFlySpeechError *)error{
    if(self.speechRecognizer != nil){
        [self.speechRecognizer cancel];
    }
    if(_channelMethod != nil && error != nil){
        NSDictionary *data = @{
                               @"code":@(error.errorCode),
                               @"desc":error.errorDesc,
                               @"type":@(error.errorType)};
        [_channelMethod invokeMethod:@"onCompleted" arguments:data];
    }
}

- (void)onResults:(NSArray *)results isLast:(BOOL)islast{
    NSMutableString *resultString = [[NSMutableString alloc] init];
    NSDictionary *dic = results[0];
    for (NSString *key in dic) {
        [resultString appendFormat:@"%@",key];
    }
    NSString * resultFromJson = [self stringFromJson:resultString];
    _result = [NSString stringWithFormat:@"%@%@",_result,resultFromJson];
    
    if(_channelMethod != nil){
        [_channelMethod invokeMethod:@"onResults" arguments:@{@"result":_result,@"isLast":@(islast)}];
    }
}

-(NSString *)stringFromJson:(NSString*)params{
    if (params == NULL) {
        return nil;
    }
    NSMutableString *tempStr = [[NSMutableString alloc] init];
    NSDictionary *resultDic  = [NSJSONSerialization JSONObjectWithData:
                                [params dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:nil];
    if (resultDic!= nil) {
        NSArray *wordArray = [resultDic objectForKey:@"ws"];
        for (int i = 0; i < [wordArray count]; i++) {
            NSDictionary *wsDic = [wordArray objectAtIndex: i];
            NSArray *cwArray = [wsDic objectForKey:@"cw"];
            for (int j = 0; j < [cwArray count]; j++) {
                NSDictionary *wDic = [cwArray objectAtIndex:j];
                NSString *str = [wDic objectForKey:@"w"];
                [tempStr appendString: str];
            }
        }
    }
    return tempStr;
}

- (void)onVolumeChanged:(int)volume{
    if(_channelMethod != nil){
        [_channelMethod invokeMethod:@"onVolumeChanged" arguments:@{@"volume":@(volume)}];
    }
}

- (void)onBeginOfSpeech{
    _result = @"";
    if(_channelMethod != nil){
        [_channelMethod invokeMethod:@"onBeginOfSpeech" arguments:@{}];
    }
}

- (void) onEndOfSpeech{
    if(_channelMethod != nil){
        [_channelMethod invokeMethod:@"onEndOfSpeech" arguments:@{@"result":_result}];
    }
}

- (void)onCancel{
    if(_channelMethod != nil){
        [_channelMethod invokeMethod:@"onCancel" arguments:@{}];
    }
}

@end

File: android/src/main/java/com/xfyun/voice/xfyunvoice/XfyunvoicePlugin.java
package com.xfyun.voice.xfyunvoice;

import android.Manifest;
import android.app.Activity;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import com.iflytek.cloud.InitListener;
import com.iflytek.cloud.RecognizerListener;
import com.iflytek.cloud.RecognizerResult;
import com.iflytek.cloud.SpeechConstant;
import com.iflytek.cloud.SpeechError;
import com.iflytek.cloud.SpeechRecognizer;
import com.iflytek.cloud.SpeechSynthesizer;
import com.iflytek.cloud.SpeechUtility;
import com.iflytek.cloud.SynthesizerListener;

import org.json.JSONObject;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;


/** XfyunvoicePlugin */
public class XfyunvoicePlugin implements MethodCallHandler {

  public final static String TAG = XfyunvoicePlugin.class.getName();

  private static Activity activity;

  private String _result = "";

  private static SpeechSynthesizer synthesizer;
  private static SpeechRecognizer recognizer;
  private static MethodChannel channel;

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    activity = registrar.activity();
    channel = new MethodChannel(registrar.messenger(), "xfyunvoice");
    channel.setMethodCallHandler(new XfyunvoicePlugin());
  }


  private void initRecognizer(){
    if(recognizer == null){
      recognizer = SpeechRecognizer.createRecognizer(activity,initListener);
    }
    recognizer.setParameter(SpeechConstant.CLOUD_GRAMMAR, null );
    recognizer.setParameter(SpeechConstant.SUBJECT, null );

    recognizer.setParameter(SpeechConstant.KEY_SPEECH_TIMEOUT,"30000");
    recognizer.setParameter(SpeechConstant.VAD_BOS,"3000");
    recognizer.setParameter(SpeechConstant.VAD_EOS,"2000");
    recognizer.setParameter(SpeechConstant.NET_TIMEOUT,"10000");
    recognizer.setParameter(SpeechConstant.SAMPLE_RATE,"16000");
    recognizer.setParameter(SpeechConstant.LANGUAGE,"zh_cn");
    recognizer.setParameter(SpeechConstant.ACCENT,"mandarin");
    recognizer.setParameter(SpeechConstant.ASR_PTT,"0");
  }

  InitListener initListener = new InitListener() {
    @Override
    public void onInit(int i) {
      Log.e(TAG,"initListener ======>onInit  i:"+i);
    }
  };

  private void initSynthesizer(){
    if(synthesizer == null){
      synthesizer = SpeechSynthesizer.createSynthesizer(activity,initListener);
    }
    SpeechUtility.getUtility().setParameter(SpeechConstant.TTS_PLAY_STATE,"tts");
    synthesizer.setParameter(SpeechConstant.SPEED,"50");
    synthesizer.setParameter(SpeechConstant.VOLUME,"50");
    synthesizer.setParameter(SpeechConstant.PITCH,"50");
    synthesizer.setParameter(SpeechConstant.SAMPLE_RATE,"16000");
    synthesizer.setParameter(SpeechConstant.VOICE_NAME,"xiaoyan");
    synthesizer.setParameter(SpeechConstant.TEXT_ENCODING,"unicode");
    synthesizer.setParameter(SpeechConstant.ENGINE_TYPE,"cloud");
  }
  SynthesizerListener syListener = new SynthesizerListener() {
    @Override
    public void onSpeakBegin() {
      _result = "";
      if(channel != null){
        Map<String,Object> data = new HashMap<>();
        channel.invokeMethod("onBeginOfSpeech",data);
      }
    }

    @Override
    public void onBufferProgress(int i, int i1, int i2, String s) {

    }

    @Override
    public void onSpeakPaused() {
    }

    @Override
    public void onSpeakResumed() {
    }

    @Override
    public void onSpeakProgress(int i, int i1, int i2) {

    }

    @Override
    public void onCompleted(SpeechError error) {
      if(channel != null){
        Map<String,Object> data = new HashMap();
        if(error != null){
          data.put("code",error.getErrorCode());
          data.put("desc",error.getErrorDescription());
        }
        data.put("type",0);
        channel.invokeMethod("onCompleted",data);
      }
    }

    @Override
    public void onEvent(int i, int i1, int i2, Bundle bundle) {

    }
  };

  RecognizerListener reListener = new RecognizerListener() {
    @Override
    public void onVolumeChanged(int i, byte[] bytes) {
      if(channel != null){
        Map<String,Integer> data = new HashMap<>();
        data.put("volume",i);
        channel.invokeMethod("onVolumeChanged",data);
      }
    }

    @Override
    public void onBeginOfSpeech() {
      if(channel != null){
        Map<String,Integer> data = new HashMap<>();
        channel.invokeMethod("onBeginOfSpeech",data);
      }
    }

    @Override
    public void onEndOfSpeech() {
      if(channel != null){ }
    }

    private HashMap<String, String> mIatResults = new LinkedHashMap<>();

    @Override
    public void onResult(RecognizerResult result, boolean b) {
      String text = JsonParser.parseIatResult(result.getResultString());
      String sn = null;
      try {
        JSONObject resultJson = new JSONObject(result.getResultString());
        sn = resultJson.optString("sn");
      } catch (Exception e) {
        e.printStackTrace();
      }
      mIatResults.put(sn, text);
      StringBuilder resBuffer = new StringBuilder();
      for (String key : mIatResults.keySet()) {
        resBuffer.append(mIatResults.get(key));
      }
      _result = resBuffer.toString();
      if(channel != null){
        Map<String,Object> data = new HashMap<>();
        data.put("result",_result);
        data.put("isLast",b);
        channel.invokeMethod("onResults",data);
        if(b){
          Map<String,String> map = new HashMap<>();
          map.put("result",_result);
          channel.invokeMethod("onEndOfSpeech",map);
        }
      }
    }

    @Override
    public void onError(SpeechError error) {
      if(channel != null){
        Map<String,Object> data = new HashMap<>();
        if(error != null){
          data.put("code",error.getErrorCode());
          data.put("desc",error.getErrorDescription());
        }
        data.put("type",0);
        channel.invokeMethod("onError",data);
      }
    }
    @Override
    public void onEvent(int i, int i1, int i2, Bundle bundle) {}
  };

  private void startListening(Result result){
    if(synthesizer != null){
      synthesizer.stopSpeaking();
    }
    if(recognizer == null){
      initRecognizer();
    }else{
      recognizer.cancel();
    }
    recognizer.setParameter("audio_source","1");
    recognizer.setParameter(SpeechConstant.RESULT_TYPE,"json");
    recognizer.setParameter(SpeechConstant.ASR_AUDIO_PATH,"asr.pcm");
    int ret = recognizer.startListening(reListener);
    Map<String,Object> data = new LinkedHashMap<>();
    data.put("ret",ret);
    data.put("msg","正在语音识别中");
    data.put("code",0);
    result.success(data);
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if("setAppId".equals(call.method)){
      try {
        String data = call.argument("args");
        SpeechUtility.createUtility(activity,data);
        result.success(true);
      } catch (Exception e) {
        e.printStackTrace();
        result.success(false);
      }
    }else if("requestPermission".equals(call.method)){
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        int permission = activity.checkSelfPermission(Manifest.permission.RECORD_AUDIO);
        if(permission != PackageManager.PERMISSION_GRANTED){
          ActivityCompat.requestPermissions(activity,new String[]{Manifest.permission.RECORD_AUDIO},101);
        }
      }
    }else if("startListening".equals(call.method)){
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        if(activity.checkSelfPermission(Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED){
          Map<String,Object> data = new LinkedHashMap<>();
          data.put("ret",false);
          data.put("msg","App未获得录音权限");
          data.put("code",-1);
          result.success(data);
        }else{
          startListening(result);
        }
      }else{
        startListening(result);
      }
    }else if("stopListening".equals(call.method)){
      if(recognizer != null){
        recognizer.stopListening();
      }
    }else if("cancel".equals(call.method)){
      if(recognizer != null){
        recognizer.cancel();
      }
    }else if("startSpeaking".equals(call.method)){
      String msg = call.arguments();
      if(synthesizer == null){
        initSynthesizer();
      }
      synthesizer.startSpeaking(msg,syListener);
    }else if("isSpeaking".equals(call.method)){
      if(synthesizer != null){
        result.success(synthesizer.isSpeaking());
      }else{
        result.success(false);
      }
    }else if("pauseSpeaking".equals(call.method)){
      if(synthesizer != null){
        synthesizer.pauseSpeaking();
      }
    }else if("resumeSpeaking".equals(call.method)){
      if(synthesizer != null){
        synthesizer.resumeSpeaking();
      }
    }else if("destroy".equals(call.method)){
      if(recognizer != null){
        recognizer.destroy();
      }
    }else {
      result.notImplemented();
    }
  }
}
